Add error dialog for critical errors on Windows

topolarity opened the pull request:
This change adds a small dialog for fatal errors on Windows:
<p align="center">
<img src="https://github.com/JuliaLang/julia/assets/84105208/9b414a16-6522-4cc1-ab3f-1cd6553f9df6" width="40%" />
</p>

with a corresponding entry in the `Application` log (in Event Viewer)
<img src="https://github.com/JuliaLang/julia/assets/84105208/19f373d4-c8ce-4596-bb36-a1229460ac71" width="70%" />


This is intended primarily for users of PackageCompiler.jl, who might be using Julia in a GUI application.

If a Julia-compiled library fails in a GUI application, users currently have no good way to know why/how their application crashed if, e.g., an exception was accidentally left unhandled in their Julia code.

Depends on https://github.com/JuliaLang/julia/pull/54835

**TODO:** Should this be a `jl_options` flag to determine whether to show the GUI? Should we log to the event log unconditionally? Is there any way that we can automatically detect whether `stderr` is connected to a terminal and user-visible?

topolarity committed:
windows: Print unhandled exceptions to `Application` log

This can be very useful when your Julia-built library hits an unexpected
exception and causes a GUI-based application to crash with no error
message.

topolarity committed:
windows: Print to Application log on critical error

fingolfin commented on src/signals-win.c:
```diff
@@ -92,9 +92,29 @@ void __cdecl crt_sig_handler(int sig, int num)
+                event_source, EVENTLOG_ERROR_TYPE, /* category */ 0, /* event_id */ (DWORD)0xE0000000L,
+               /* user_sid */ NULL, /* n_strings */ 1, /* data_size */ 0, strings, /* data */ NULL
+            );
+            MessageBoxW(NULL, /* message */ L"error: libjulia received a fatal signal.\n\n"
```
I know very little about Windows, so my apologies if these questions are dumb, but: 1. Is this call async, or does it block? 
2. Is this really safe to call here? In particular, 
3. I am slightly worried about situations I experienced in an entirely different context were broadly similar code could lead to hundreds of dialog boxes spamming a user (when code that launched a ton of subprocesses to process data had them all fail simultaneously -- e.g. because a network connection dropped).

Anyway, none of this is meant as argument against this PR (which to me, as a non-Windows user, seems quite useful). Feel free to ignore me :-)

davidanthoff replied:
On 1: it blocks the thread.

topolarity replied:
a151cf3 added a `jl_options` guard

It's disabled by default now, but will be enabled by PackageCompiler when exporting Julia-based libraries - hopefully that resolves some of the concern. These are "critical" errors, so in some sense they should never happen (although in practice aggressive `Ctrl-C` at the REPL makes them happen quite a bit, so it's good that we don't alert in that case)

davidanthoff commented:
Probably also needs to detect scenarios where no windows desktop is present? SSH, PowerShell remoting, running as a service etc. 

Somehow making the dialog box optional seems like a good idea to me.

vtjnash commented on src/signals-win.c:
```diff
@@ -278,59 +298,88 @@ LONG WINAPI jl_exception_handler(struct _EXCEPTION_POINTERS *ExceptionInfo)
         }
     }
+    ios_t full_error, summary;
+    ios_mem(&full_error, 1024);
```
calling malloc unconditionally here is a bit wasteful, since we will call it on demand when actually needed
```suggestion
    ios_mem(&full_error, 0);
```

vtjnash commented on src/support/ios.c:
```diff
@@ -988,6 +988,24 @@ ios_t *ios_file(ios_t *s, const char *fname, int rd, int wr, int create, int tru
 
+#ifdef _OS_WINDOWS_
+const wchar_t *ios_utf8_to_wchar(const char *str) {
+    /* Fast-path empty strings, as MultiByteToWideChar() returns zero for them. */
+    if (str[0] == '\0') {
+        wchar_t *wstr = (wchar_t *)malloc(sizeof(wchar_t));
+        wstr[0] = L'\0';
+        return wstr;
+    }
+    size_t len = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL, 0);
+    if (!len)
+        return NULL;
+    wchar_t *wstr = (wchar_t *)malloc(len * sizeof(wchar_t));
+    if (!wstr || !MultiByteToWideChar(CP_UTF8, 0, str, -1, wstr, len))
+        return NULL;
+    return wstr;
```
```suggestion
    /* Fast-path empty strings, as MultiByteToWideChar() returns zero for them. */
    ssize_t wlen = uv_wtf8_length_as_utf16(str);
    wchar_t *wstr = (wchar_t *)malloc_s(sizeof(wchar_t) * wlen);
    uv_wtf8_to_utf16(str, wstr, wlen);
    return wstr;
```
This is a more reliable (and concise) version of MultiByteToWideChar now provided by libuv

vtjnash commented on src/signals-win.c:
```diff
@@ -278,59 +298,88 @@ LONG WINAPI jl_exception_handler(struct _EXCEPTION_POINTERS *ExceptionInfo)
+    if (event_source != INVALID_HANDLE_VALUE) {
+        ios_putc('\0', &full_error);
+        const wchar_t *strings[] = {
+            ios_utf8_to_wchar(full_error.buf),
```
We should free this

topolarity replied:
Good catch, thank you.

vtjnash reviewed (commented):
SGTM, I think this should work okay

DilumAluthge commented:
Was this closed intentionally?

topolarity commented:
> Was this closed intentionally?

It was, but https://github.com/JuliaLang/julia/pull/54835 has since been re-opened

I'll be re-visiting this soon since "Julia as a library" scenarios on Windows are becoming relevant soon

topolarity committed:
Add `jl_options.alert_on_critical_error` to control alert dialog

This is disabled by default, but it can be enabled by downstream
applications (in particular, by JuliaC.jl / PackageCompiler) so
Windows users with a Julia library in a pure GUI application will
get a useful error message for a crash.

vtjnash commented on src/task.c:
```diff
@@ -732,11 +732,36 @@ JL_DLLEXPORT JL_NORETURN void jl_no_exc_handler(jl_value_t *e, jl_task_t *ct)
+    HANDLE event_source = RegisterEventSourceW(NULL, L"julia");
+    if (event_source != INVALID_HANDLE_VALUE) {
+        ios_putc('\0', &s);
+        const wchar_t *strings[] = { ios_utf8_to_wchar(s.buf) };
```
Free this?

topolarity replied:
Sorry for rushing the rebase - This should be covered now.

vtjnash commented:
build failure looks related```
C:/workdir/src/support/ios.c:995:5: error: implicit declaration of function 'uv_wtf8_to_utf16' [-Wimplicit-function-declaration]
2025-09-19 20:40:54 EDT
  995 |     uv_wtf8_to_utf16(str, wstr, wlen);
```

topolarity commented:
> build failure looks related
> 
> ```
> C:/workdir/src/support/ios.c:995:5: error: implicit declaration of function 'uv_wtf8_to_utf16' [-Wimplicit-function-declaration]
> 2025-09-19 20:40:54 EDT
>   995 |     uv_wtf8_to_utf16(str, wstr, wlen);
> ```

Fixed.

Do we have not have `-Werror=implicit-function-declaration` configured?
